### PR TITLE
Rdcc 452

### PR DESCRIPTION
**Pull request by ConorH994**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-452

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -188,6 +188,7 @@ pitest {
     outputFormats = ['XML', 'HTML']
     timestampedReports = false
     mutationThreshold = 90
+    useClasspathFile = true
```
Comment (AdilOoze) [inline build.gradle]: Remove this Conor 👎

File: src/integrationTest/java/uk/gov/hmcts/reform/userprofileapi/integration/CreateNewUserProfileWithDuplicateUserIntTest.java [hunk 2]
```diff
@@ -170,7 +168,7 @@ public void should_return_201_and_create_user_profile_when_duplicate_in_sidam()
     @Test
     public void should_return_201_and_create_user_profile_when_status_not_properly_returned_by_sidam() throws Exception {
 
-        mockWithGetSuccess(true);
+        mockWithGetSuccess(false);
```
Comment (AdilOoze) [inline src/integrationTest/java/uk/gov/hmcts/reform/userprofileapi/integration/CreateNewUserProfileWithDuplicateUserIntTest.java]: is this meant to be changed?

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/client/GetUserProfileWithRolesResponseTest.java [hunk 3]
```diff
@@ -13,7 +13,6 @@
     @Test
     public void should_hold_values_after_creation_when_user_roles_required() {
 
-
```
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/userprofileapi/client/GetUserProfileWithRolesResponseTest.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java [hunk 4]
```diff
@@ -2,10 +2,13 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.*;
```
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java]: ?

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java [hunk 5]
```diff
@@ -28,9 +32,38 @@
...
+        userProfileService.update(updateUserProfileData, "1234");
+
+        UserProfileRolesResponse userProfileRolesResponse = mock(UserProfileRolesResponse.class);
+
+        when(resourceUpdator.updateRoles(updateUserProfileData, "1234")).thenReturn(userProfileRolesResponse);
+        userProfileRolesResponse = userProfileService.updateRoles(updateUserProfileData, "1234");
+
+        assertThat(userProfileRolesResponse).isNotNull();
+
+        //verify(userProfileService, times (1)).updateRoles(any(), "1234");
```
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java]: Delete this

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java [hunk 6]
```diff
@@ -28,9 +32,38 @@
...
+        //verify(userProfileService, times (1)).update(updateUserProfileDataMock, "1234");
```
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java]: and this

File: src/integrationTest/java/uk/gov/hmcts/reform/userprofileapi/integration/CreateNewUserProfileWithDuplicateUserIntTest.java (hunk 2 above)
Comment (ConorH994) [reply src/integrationTest/java/uk/gov/hmcts/reform/userprofileapi/integration/CreateNewUserProfileWithDuplicateUserIntTest.java]: yes

File: src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java (hunk 4 above)
Comment (ConorH994) [reply src/test/java/uk/gov/hmcts/reform/userprofileapi/domain/service/UserProfileServiceTest.java]: done

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java [hunk 7]
```diff
@@ -265,12 +259,22 @@
...
+    ) {
         log.info("Updating user profile");
+        UserProfileRolesResponse userProfileResponse = new UserProfileRolesResponse();
+        if (StringUtils.isEmpty(updateUserProfileData.getRolesAdd())) {
 
-        userProfileService.update(updateUserProfileData, userId);
-        return ResponseEntity.status(HttpStatus.OK).build();
+            log.info("Updating user profile without roles");
+            userProfileService.update(updateUserProfileData, userId);
+        } else {
```
Comment (akiohmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: Why are you using a validator but not returning a boolean? It looks like validateUserProfileDataAndUserId is doing more than validating and manipulating the data somehow. Please rename.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java [hunk 8]
```diff
@@ -265,12 +259,22 @@
             consumes = APPLICATION_JSON_UTF8_VALUE,
             produces = APPLICATION_JSON_UTF8_VALUE
     )
+
     @ResponseBody
-    public ResponseEntity<CreateUserProfileResponse> updateUserProfile(@Valid @RequestBody UpdateUserProfileData updateUserProfileData, @PathVariable String userId) {
+    public ResponseEntity<UserProfileRolesResponse> updateUserProfile(@Valid @RequestBody UpdateUserProfileData updateUserProfileData, @PathVariable String userId
+    ) {
```
Comment (akiohmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: Super hard to read. This is very hard-to-read and confusing code. Consider refactoring.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java [hunk 9]
```diff
@@ -28,42 +41,91 @@
...
+            status = HttpStatus.INTERNAL_SERVER_ERROR;
+        }
+        persistAudit(status, userProfile, ResponseSource.SYNC);
         return userProfile;
     }
 
-    public void persistAudit(HttpStatus idamStatus, UserProfile userProfile) {
-        Audit audit = new Audit(idamStatus.value(), resolveStatusAndReturnMessage(idamStatus), ResponseSource.SYNC, userProfile);
+    @Override
+    public UserProfileRolesResponse updateRoles(UpdateUserProfileData profileData, String userId) {
```
Comment (akiohmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java]: This is very hard-to-read and confusing code. Consider refactoring.

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java (hunk 7 above)
Comment (ConorH994) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: The validator just throws a missing field exception if there is invalid data in the request. There doesn't seem to be any reason to return a boolean

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java (hunk 9 above)
Comment (ConorH994) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/UserProfileUpdator.java]: Removed multi return statement

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java (hunk 8 above)
Comment (ConorH994) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: Removed @Valid annotation from method declaration

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java (hunk 7 above)
Comment (akiohmcts) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/controller/UserProfileController.java]: See below comment. You need to try to code in a conventional way, which means making it easy for the reviewer to understand what it is you are doing. So either return a boolean and explicitly throw a Runtime exception preferably in the same layer where the exception is occuring. See example below.
If you have a lot of validation logic wrapped in a validation object, fine. But flat is better than nested, so doing something really critical like throwing a runtime exception needs to be made explicit and not obfuscated deep inside some validation logic with a poorly named method that gives me no indication that any exception is being thrown.

**Pull request merged**